astropy/nddata/covariance.py

```diff
@@ -0,0 +1,1034 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Defines a class used to store and interface with covariance matrices.
+"""
+
+import warnings
+from pathlib import Path
+
+import numpy as np
+from matplotlib import pyplot as plt
+from scipy import sparse
+
+from astropy import log
+from astropy.io import fits
+
+__all__ = ["Covariance"]
+
+
+class Covariance:
+    r"""
+    A general utility for storing, manipulating, and file I/O of sparse
+    covariance matrices.
+
+    .. warning::
+
+        Although possible to abstract further, use of ``raw_shape``
+        can only be 2D in the current implementation.
+
+    Parameters
+    ----------
+    inp : `scipy.sparse.csr_matrix`, optional
```

Comment:
`scipy` is still an "optional" dependency for `astropy`. Does this mean anything touching the stuff from this module requires `scipy` to run?

Reply:
Please put your thoughts here:

* #12448

Reply:
See my review comments; I think if it is generally only available with scipy, that's acceptable.
Whether it would be worth offering a stripped-down version with dense numpy arrays depends on the extra work involved with that, and if you'd still consider such an option usable given the performance impact.
There are certainly plenty of use cases with array lengths of only ~100 or less where it probably would not hurt, as in the original `specutils` example where this came up. But then `specutils` already depends on scipy, so would never need to use that alternative version.